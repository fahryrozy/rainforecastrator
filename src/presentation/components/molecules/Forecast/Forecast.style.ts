import {scale} from '@core/style';
import {StyleSheet} from 'react-native';

export const styles = StyleSheet.create({
  forecastCard: {
    borderRadius: scale(10),
    marginVertical: scale(5),
  },
  title: {
    display: 'flex',
    height: scale(45),
    backgroundColor: '#222',
    opacity: 0.6,
    borderTopLeftRadius: scale(10),
    borderTopRightRadius: scale(10),
    marginVertical: scale(1),
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: scale(5),
    paddingHorizontal: scale(15),
  },
  content: {
    height: scale(145),
    backgroundColor: '#222',
    opacity: 0.6,
    borderBottomRadius: scale(10),
    flexDirection: 'row',
    borderBottomLeftRadius: scale(10),
    borderBottomRightRadius: scale(10),
    paddingHorizontal: scale(10),
  },
  renderedCard: {
    width: 80,
    padding: 5,
    justifyContent: 'center',
    alignItems: 'center',
  },
  icon: {
    width: scale(60),
    height: scale(60),
  },
  temp: {
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
    height: '70%',
    backgroundColor: '#0bd',
    textAlign: 'center',
    color: 'rgba(255, 255, 255, 1)',
    fontFamily: 'Poppins',
    fontSize: scale(64),
  },
  tempText: {
    textAlign: 'center',
    color: 'rgba(255, 255, 255, 1)',
    fontFamily: 'Poppins',
    fontSize: scale(64),
  },
  loc: {
    rowGap: scale(10),
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
    height: '30%',
    backgroundColor: '#000',
  },
  locText: {
    textAlign: 'center',
    color: 'rgba(255, 255, 255, 1)',
    fontFamily: 'Poppins',
    fontSize: scale(24),
    fontWeight: '400',
  },
  minMaxText: {
    textAlign: 'center',
    color: 'rgba(255, 255, 255, 1)',
    fontFamily: 'Poppins',
    fontSize: scale(20),
    fontWeight: '400',
  },
});
