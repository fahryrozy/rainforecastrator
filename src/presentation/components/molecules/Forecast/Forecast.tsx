import {
  View,
  Text,
  FlatList,
  Image,
  TouchableOpacity,
  ActivityIndicator,
} from 'react-native';
import React from 'react';
import moment from 'moment';
import {styles} from './Forecast.style';
import {Hour} from '@domain/entities';
import {Typhography} from '@core/style';

type ItemProps = {hourly_forecast: Hour; onPress: any; isLoading: boolean};

const RenderedCard: React.FC<ItemProps> = ({
  hourly_forecast,
  onPress,
  isLoading,
}) => (
  <>
    {isLoading ? (
      <View style={styles.renderedCard}>
        <ActivityIndicator size="small" color="#fff" />
      </View>
    ) : (
      <TouchableOpacity style={styles.renderedCard} onPress={onPress}>
        <Text
          style={styles.tempText}
          testID="fc-temp">{`${hourly_forecast.temp_c}Â°`}</Text>
        <Image
          source={{uri: `https:${hourly_forecast.condition.icon}`}}
          style={styles.icon}
          testID="fc-img"
        />
        <Text style={styles.timeText} testID="fc-time">{`${moment(
          hourly_forecast.time,
        ).format('h A')}`}</Text>
      </TouchableOpacity>
    )}
  </>
);

type Props = {
  data: Array<Hour>;
  date: string;
  setCurrCondition: (item: Hour) => void;
  isToday: boolean;
  isLoading: boolean;
};

const Forecast: React.FC<Props> = ({
  data,
  date,
  setCurrCondition,
  isToday,
  isLoading,
}) => {
  moment.locale('en');
  return (
    <View style={styles.forecastCard}>
      <View style={styles.title}>
        {isToday ? (
          <Text style={Typhography.headerDefault()}>
            Today's hourly forecast
          </Text>
        ) : (
          <Text style={Typhography.headerDefault()}>
            {moment(date).calendar({
              sameDay: '[Today]',
              nextDay: '[Tomorrow]',
              nextWeek: 'dddd',
              lastDay: '[Yesterday]',
              lastWeek: '[Last] dddd',
              sameElse: 'DD/MM/YYYY',
            })}
            's hourly forecast
          </Text>
        )}
      </View>
      <View style={styles.content}>
        {data && (
          <FlatList
            horizontal={true}
            data={data}
            renderItem={({item}) => (
              <RenderedCard
                hourly_forecast={item}
                isLoading={isLoading}
                onPress={() => setCurrCondition(item)}
              />
            )}
            showsHorizontalScrollIndicator={false}
          />
        )}
      </View>
    </View>
  );
};
export default Forecast;
