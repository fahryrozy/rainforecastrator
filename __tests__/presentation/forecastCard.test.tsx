/**
 * @format
 */

import 'react-native';
import React from 'react';
import Forecast from '@presentation/components/molecules/Forecast/Forecast';
import {it} from '@jest/globals';
import renderer from 'react-test-renderer';

describe('Forecast Card in Homescreen view which used to see forecasting data', () => {
  it('renders correctly with blank data', () => {
    const isToday = true;
    const isLoading = false;
    const setCurrCondition = jest.fn();
    const date = '1900-01-01';
    renderer.create(
      <Forecast
        data={[]}
        setCurrCondition={setCurrCondition}
        date={date}
        isToday={isToday}
        isLoading={isLoading}
      />,
    );
  });

  it('renders correctly with mock data', () => {
    const isToday = true;
    const isLoading = false;
    const setCurrCondition = jest.fn();
    const date = '1900-01-01';
    renderer.create(
      <Forecast
        data={[]}
        setCurrCondition={setCurrCondition}
        date={date}
        isToday={isToday}
        isLoading={isLoading}
      />,
    );
  });

  it('match snapshot', () => {
    const isToday = true;
    const isLoading = false;
    const setCurrCondition = jest.fn();
    const date = '1900-01-01';
    const snap = renderer
      .create(
        <Forecast
          data={[]}
          setCurrCondition={setCurrCondition}
          date={date}
          isToday={isToday}
          isLoading={isLoading}
        />,
      )
      .toJSON();
    expect(snap).toMatchSnapshot();
  });

  it('has hourly forecast title text', () => {});
  it('has forecast weather temp', () => {});
  it('has forecast weather image', () => {});
  it('has forecast weather time', () => {});
});
