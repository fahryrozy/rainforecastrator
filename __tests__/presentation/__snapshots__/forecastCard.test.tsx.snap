// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forecast Card in Homescreen view which used to see forecasting data match snapshot 1`] = `
<View
  style={
    {
      "borderRadius": 20,
      "marginVertical": 10,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#222",
        "borderTopLeftRadius": 20,
        "borderTopRightRadius": 20,
        "display": "flex",
        "flexDirection": "row",
        "height": 90,
        "marginVertical": 2,
        "opacity": 0.6,
        "paddingHorizontal": 30,
        "paddingVertical": 10,
      }
    }
  >
    <Text
      style={
        {
          "color": "#fff",
          "fontFamily": "Quicksand-Bold",
          "fontSize": 36,
        }
      }
    >
      Today's hourly forecast
    </Text>
  </View>
  <View
    style={
      {
        "backgroundColor": "#222",
        "borderBottomLeftRadius": 20,
        "borderBottomRadius": 20,
        "borderBottomRightRadius": 20,
        "flexDirection": "row",
        "height": 290,
        "opacity": 0.6,
        "paddingHorizontal": 20,
      }
    }
  >
    <RCTScrollView
      data={[]}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={true}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <View />
    </RCTScrollView>
  </View>
</View>
`;
